comictalker/comictalker.py

```diff
@@ -0,0 +1,200 @@
+"""Handles collecting data from source talkers.
+"""
+# Copyright 2012-2014 Anthony Beville
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+import logging
+from typing import Callable
+
+from comicapi.genericmetadata import GenericMetadata
+from comictalker.resulttypes import ComicIssue, ComicVolume
+from comictalker.talkerbase import SourceStaticOptions, TalkerError
+
+logger = logging.getLogger(__name__)
+
+
+# To signal image loaded etc.
+def list_fetch_complete(url_list: list[str]) -> None:
+    ...
+
+
+def url_fetch_complete(image_url: str, thumb_url: str | None) -> None:
+    ...
+
+
+class ComicTalker:
```

Comment:
I kind of follow but then also don't. It maybe because I'm thinking ahead to multiple talkers. If I understand correctly, having the functions "passed through" (proxied) via `series_result = self.talker.search_for_series(series_name, callback, refresh_cache, literal)` in `comictalker` for example is the issue? Why not have it use CV directly? If so, here's my thinking:

- Looking ahead, if multiple talkers were to be queried, it'd be `comictalker` that would aggregate the results together.
- Running some post data manipulation would be done here too. Off the top of my head now, the only example is writing the metadata notes fields. Should that be done by the talker or `comictalker`?
- Right now, only NotImplemented errors are covered but that should probably be expanded. Should `comictalker` be the one that does the retries?
- How would NotImplemented be handled, in talkerbase?

I might have the wrong end do of the stick here though?

With get_talkers that will do more in the future as was demoed with the example code to load all sources in the source dir. Doing some basic checks that a talker has the required options etc. should be in there.

Reply:
You bring up some good points

1. It wouldn't make sense to aggregate them inside ComicTalker because it's purpose is to retrieve data not make decisions on which data should be returned or merged, and if by aggregate you mean just returning a list of results that's as simple as:
   ```python
   results = []
   for talker in talker_list:
     results.append(talker.Search_for_series(series_name, callback, refresh_cache, literal))
   # results now contains all of the results for all of the wanted talkers
   ```
1. Post data manipulation also wouldn't be done here, ComicTalker currently only retrieves and assembles data the one choice it does make is how many results to return when searching, every other decision such as whether or not to even display the result is handled in `comictaggerlib`. In regards to the notes that should be handled by `comictaggerlib` because this is a separate package that serves the purpose of retrieving data, comicapi still has a TODO from 2015 when it was split out of `comictaggerlib` because the CBI format records the application that created/edited the file but because it was split out and stands on it's own it no longer has access to (or shouldn't) `comictaggerlib`
https://github.com/comictagger/comictagger/blob/b2532ce03a904fa0a44c4e9fd463d8b3e56d79ef/comicapi/comicbookinfo.py#L106-L108
https://github.com/comictagger/comictagger/blob/2daf9b3ed884cbb2c4b3884f863992e4666a4b8a/comicapi/comicbookinfo.py#L139-L145
1. `NotImplemented` Exceptions are actually not handled, it's logged that an error happened but then another Exception is raised, that doesn't change anything about how a consumer of this class would operate, they would still have to catch an Exception and decide what to do, now instead of catching `NotImplemented` they catch `TalkerError` and separately check what that error is. To handle it, it would have to return something that could be used, so for searching, it would be possible to return an empty list but then comictaggerlib (or the user using ComicTagger because comictaggerlib wouldn't have a way to tell them) wouldn't know that the way that was searched wasn't implemented, it would just be a search that returned nothing.
1. `NotImplemented` would be handled in `comictaggerlib` where Exceptions are currently handled.

Reply:
Is this in the right area https://github.com/comictagger/comictagger/pull/341/commits/cab69a32be8dcc3b91bec8caeec1fca18c1fd1f6 ? I'm wondering about moving `static_options` and `settings_options` onto the root and removing `source_details`. It's tidier but I'm not sure it's giving any real value?

Reply:
No I meant this:
```python
"""Handles collecting data from source talkers."""

# Copyright 2012-2014 Anthony Beville
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from __future__ import annotations

import logging
from typing import Callable

import comictalker.talkers.comicvine
from comictalker.talkerbase import TalkerBase, TalkerError

logger = logging.getLogger(__name__)


def get_comic_talker(source_name: str, log_func: Callable[[str], None]) -> TalkerBase:
    # Retrieve the available sources modules
    sources = get_talkers()
    if source_name not in sources:
        raise TalkerError(source=source_name, code=4, desc="The talker does not exist")

    talker = sources[source_name]
    talker.log_func = log_func
    return talker


def get_talkers():
    return {"comicvine": comictalker.talkers.comicvine.ComicVineTalker()}
```

Reply:
I was way off then...

Just to make sure I've got this correct, `main` would now be this?
```
# talker_api = ComicTalker(settings.comic_info_source) <- Current
talker_api = nametbd.get_comic_talker(settings.comic_info_source)
```
Thinking out loud here. Would the idea be for generating the settings menu for each talker to use `get_talkers` (`temp_list_of_takers = nametbd.get_talkers()`) and then delete the dict of objects after? At the moment the comictalker object has them all. Outside of generating the settings menus, there's no reason to have them live?

Reply:
Yes, that is what main would be.
Yes on generating the settings, though there is one more change which would make sense as well is returning the class instead of the object, this allows us to pass the settings in directly instead of re-reading the settings file (don't actually worry about changing the settings code now, we'll figure that out later down the line).
instead of:
```python
def get_talkers():
    return {"comicvine": comictalker.talkers.comicvine.ComicVineTalker()}

talker_api = nametbd.get_comic_talker(settings.comic_info_source)
```
we do
```python
def get_talkers():
    return {"comicvine": comictalker.talkers.comicvine.ComicVineTalker}

cv_api = nametbd.get_comic_talker(settings.comic_info_source)
talker_api = cv_api()
# or
talker_api = nametbd.get_comic_talker(settings.comic_info_source)()
```

Reply:
After that long and winding road how is https://github.com/comictagger/comictagger/pull/341/commits/561dc2804445f1d49da73446c3c7f810486c5938 ? :)

Reply:
Still need to remove the dregs of `fetch_alternate_cover_urls` but other than that this conversation is done.
(I keep changing my mind on whether or not `fetch_comic_data` should be split into multiple functions or not, but at this point it doesn't actually change anything)

Reply:
I've not dug into this yet but I'm not sure that anything is needed other than fetch by id. CLI calls with volume id and issue number but I would expect if it knows those two things it should know the issue id as well? Otherwise the user has clicked and so we should have an issue id too? There is this code in CV:
```
    def fetch_issue_data(self, series_id: int, issue_number: str) -> GenericMetadata:
        issues_list_results = self.fetch_issues_by_volume(series_id)

        # TODO Is this required?
        f_record = None
        for record in issues_list_results:
            if not IssueString(issue_number).as_string():
                issue_number = "1"
            if (
                IssueString(record["issue_number"]).as_string().casefold()
                == IssueString(issue_number).as_string().casefold()
            ):
                f_record = record
                break
```
That seems to be searching through all the issues (again after the issue has been identified?) and is adding issue "1" if there is no issue but there always should be? I've not looked into it properly but maybe you know off the top of your head why it needs to do that and doesn't have the issue id? Otherwise, I'll take a look at it.

Reply:
I couldn't find a reason so I made these changes https://github.com/comictagger/comictagger/pull/341/commits/4f42fef4fc3ef3cec2f94c68e05e14978288e204 See if I've missed something? I've left a cover in for volume id and issue number but that should not be needed?

If it holds true that the issue_id (or series_id in the case of MU etc.) is always available then `fetch_comic_data` could be reduced to a single var.

Reply:
These changes look good, there was probably never a requirement just what made sense at he time.
Yes this is true that we could probably do with only one but let's leave it as is for now, and you didn't miss anything everything uses the issue_id.